src/main/scala/viper/gobra/translator/encodings/combinators/TypeEncoding.scala

```diff
@@ -121,7 +122,7 @@ trait TypeEncoding extends Generator {
       for {
         footprint <- addressFootprint(ctx)(loc, in.FullPerm(loc.info))
         eq1 <- ctx.equal(loc, in.DfltVal(t.withAddressability(Exclusive))(loc.info))(loc)
-        eq2 <- ctx.equal(in.Ref(loc)(loc.info), in.NilLit(in.PointerT(t, Exclusive))(loc.info))(loc)
+        eq2 <- ctx.equal(in.UncheckedRef(loc)(loc.info), in.NilLit(in.PointerT(t, Exclusive))(loc.info))(loc)
```

Comment:
Can you elaborate on why you picked `UncheckedRef` here instead of `Ref`?

Reply:
The way we model memory, taking a reference never fails (as it is a pure operation). Conversely, taking a reference in Go fails if the argument is nil. This mismatch caused the unsoundness. To fix the mismatch, I added the (Safe)Ref node, which artificially fails if the argument is nil. Whenever there is an actual reference-taking happening in the Go program, we will have to use `Ref`. However, in many cases, such as here, we take a reference in the encoding even though there is no reference-taking in Go. In these cases, we can use `UncheckedRef`. In such cases, we use references to modularly split the encoding.